File: library/src/main/java/io/github/pelmenstar1/rangecalendar/YearMonth.java
package io.github.pelmenstar1.rangecalendar;

final class YearMonth {
    public static int create(int year, int month) {
        return (year * 12) + (month - 1);
    }

    public static int forDate(int date) {
        return create(DateInt.getYear(date), DateInt.getMonth(date));
    }

    public static int getYear(int ym) {
        return ym / 12;
    }

    public static int getMonth(int ym) {
        return (ym % 12) + 1;
    }
}

File: library/src/main/java/io/github/pelmenstar1/rangecalendar/YearMonth.kt
@file:Suppress("NOTHING_TO_INLINE")

package io.github.pelmenstar1.rangecalendar

internal fun YearMonth(year: Int, month: Int): YearMonth {
    return YearMonth(year * 12 + (month - 1))
}

@JvmInline
internal value class YearMonth(val totalMonths: Int) {
    val year: Int
        get() = totalMonths / 12

    val month: Int
        get() = totalMonths % 12 + 1

    inline operator fun component1() = year
    inline operator fun component2() = month

    operator fun minus(other: YearMonth): YearMonth {
        return YearMonth(totalMonths - other.totalMonths)
    }

    operator fun plus(months: Int): YearMonth {
        return YearMonth(totalMonths + months)
    }

    operator fun minus(months: Int): YearMonth {
        return YearMonth(totalMonths - months)
    }

    companion object {
        fun forDate(date: PackedDate) = YearMonth(date.year, date.month)
    }
}